File: 113.R
# Bouncy numbers

library(memoise)

options(scipen=100)

Args     <- commandArgs()
exponent <- ifelse(is.na(Args[6]), 100, as.numeric(Args[6]))

count_increasing <- memoise(function(last_digit, len) {
    if ( len == 1 ) {
        return(1)
    }

    count <- 0
    for ( prev_last_digit in seq.int(last_digit) ) {
        count <- count + count_increasing(prev_last_digit, len - 1)
    }

    return(count)
})

count_decreasing <- memoise(function(last_digit, len) {
    if ( len == 1 ) {
        return(1)
    }

    count <- 0
    for ( prev_last_digit in seq.int(last_digit, 9) ) {
        count <- count + count_decreasing(prev_last_digit, len - 1)
    }

    return(count)
})

increasing <- 0
for ( num_digits in seq.int(exponent) ) {
    for ( last_digit in seq.int(9) ) {
        increasing <- increasing + count_increasing(last_digit, num_digits)
    }
}

decreasing <- 0
for ( num_digits in seq.int(exponent) ) {
    for ( last_digit in seq.int(0, 9) ) {
        decreasing <- decreasing + count_decreasing(last_digit, num_digits)
    }
    decreasing <- decreasing - 1
}

double_count <- 9 * exponent

cat(increasing + decreasing - double_count, fill=TRUE)

File: 113.js
#!/usr/bin/env node

// Non-bouncy numbers

var program = require('commander');

program
    .version('0.1.0')
    .description('Non-bouncy numbers')
    .option('-e, --exponent <int>', 'Maximum value of n as an exponent of 10',
        Number, 100)
    .parse(process.argv);

var count_increasing = (function() {
    var memo = {};

    return function(last_digit, length) {
        var key = last_digit.toString() + ':' + length.toString();

        if (key in memo) {
            return memo[key];
        }

        if (length == 1) {
            memo[key] = 1;
            return 1;
        }

        var count = 0;
        for (var prev = 1; prev <= last_digit; prev++) {
            count += count_increasing(prev, length - 1);
        }

        memo[key] = count;
        return count;
    };
})();

var count_decreasing = (function() {
    var memo = {};

    return function(last_digit, length) {
        var key = last_digit.toString() + ':' + length.toString();

        if (key in memo) {
            return memo[key];
        }

        if (length == 1) {
            memo[key] = 1;
            return 1;
        }

        var count = 0;
        for (var prev = last_digit; prev <= 9; prev++) {
            count += count_decreasing(prev, length - 1);
        }

        memo[key] = count;
        return count;
    };
})();

var increasing = 0;
for (var num_digits = 1; num_digits <= program.exponent; num_digits++) {
    for (var last_digit = 1; last_digit <= 9; last_digit++) {
        increasing += count_increasing(last_digit, num_digits);
    }
}

var decreasing = 0;
for (var num_digits = 1; num_digits <= program.exponent; num_digits++) {
    for (var last_digit = 0; last_digit <= 9; last_digit++) {
        decreasing += count_decreasing(last_digit, num_digits);
    }
    decreasing--;
}

var double_count = 9 * program.exponent;

console.log(increasing + decreasing - double_count);
